### ISSUE TITLE
Questions about using MLMG solver to solve the elliptic equation with inflow boundary condition

### ISSUE THREAD

**Issue by ruohai0925** [AMReX-Codes/amrex#258]
Hello all,

It seems that the Cell centered MLMG solver does not support the inflow boundary condition.  When I use the MLMG solver, it will go into the function MLMGBndry::setBoxBC, but this function does not support the LinOpBCType::inflow. 

See here: https://github.com/AMReX-Codes/amrex/blob/5553b6af63da745ef13ecdb51654b72ec1cca544/Src/LinearSolvers/MLMG/AMReX_MLMGBndry.cpp#L60

Can someone give me some help? Will this solver be improved?

Thanks so much.

(Better to use my English name "Jordan" instead my nick name "Ruohai" to ask questions later. )

Jordan

**Comment by asalmgren** [comment]
Jordan,

You are mis-understanding the LinOpBCType.

The two boundary conditions that are supported are Dirichlet (the value
itself is set) and Neumann (the normal derivative is set).

If you are solving a diffusion equation for phi, for example, and want to
set the value of phi at inflow, then that is a Dirichlet boundary
condition. If you are solving a fluid equation for pressure and inflow has
zero pressure gradient, then you would use the Neumann boundary condition.

**Comment by ruihai0925** [comment]
Thanks, Ann

I think I am misunderstanding the differences between the physical boundary condition and mathematical boundary condition used for the solver. 

Another additional question is about the location of bc values. What I learned is for physical bc, they are located on the face. For internal bc (including the periodic bc), no matter amr is considered or not, they are located on the cell-center (things like 0.5*dx).

Here is an example for the lid-driven case:

![image](https://user-images.githubusercontent.com/30739800/39784212-43f23746-52dd-11e8-8901-2fbfb0bd3568.png)

So from the coding aspect, we set u(i,j+1)=1. But the package will know that u(i,j+1) is defined at the red star, not the blue circle point, which means if I want to calculate the du/dy at green point, we need to be very careful the location of the red points.

Not sure if I understand right. Thanks for helping to verify it.

Jordan

**Comment by drummerdoc** [comment]
Yes.  You are correct here.  The Dirichlet data is stored in a container that happens to be cell-centered but has the correct number of slots to hold the data.  It certainly can be confusing!

**Comment by drummerdoc** [comment]
To be complete (and risk being more confusing), there is another situation where things are funny.  For the cell-centered multi-level linear solves, the Dirichlet data along the coarse-fine interface actually lives dx_coarse/2 away from the interface.  It is interpolated tangential to the c-f boundary using only coarse data.  The linear operators that use this data "know" where physical boundary and coarse-fine boundary data live.  To do their work, they typically will construct a cell-centered intermediate value based on the Dirichlet data and the internal solution, and then the stencil operation can proceed without knowing about one-sided or shifted stencils.

**Comment by ruihai0925** [comment]
Thanks so much Marc. 

Yeah, after doing the parallel interpolation (the Dirichlet data lives dx_coarse/2 away from the interface), applyBC is used for perpendicular interpolation to the dx_fine/2. I asked this question several months ago (https://github.com/AMReX-Codes/amrex/issues/147). Now I have validated the MLMG solver in multi-levels. Cool!

Jordan

**Comment by ruihai0925** [comment]
Hello all, 

Thanks so much for your previous help.

I think now I have finished validating the MLMG solver by some cases. **The question now is how to calculate multilevel cell centered rhs from multilevel cell centered vel.** I have written a naive version which can deal with the periodic boundary condition. The difficulty lies in that some ghost cells for velocity are at the domain edge.  So one has to use the mask technique (https://github.com/AMReX-Codes/IAMR/issues/20) to consider different situations when calculating the rhs. 

Then I compared ML_Cell_LinOp.H and ML_Node_LinOp. The latter provides the compRHS function as an interface, that is what IAMR uses. Maybe I need to rewrite them on my own by using tricks in the Dr DF.Martin's papers, 

**Not quite sure if compRHS will be implemented into the Cell_LinOp recently.** Any suggestions are appreciated.

Thanks.

References:
[1] Martin D F, Colella P. A cell-centered adaptive projection method for the incompressible Euler equations[J]. Journal of computational Physics, 2000, 163(2): 271-312.

[2] Martin D F, Colella P, Graves D. A cell-centered adaptive projection method for the incompressible Navier–Stokes equations in three dimensions[J]. Journal of Computational Physics, 2008, 227(3): 1863-1886.
